crates/ruff/src/jupyter/notebook.rs

```diff
@@ -130,60 +173,201 @@ impl JupyterNotebook {
+    /// where we need to check for the trailing newline.
+    ///
+    /// The index building is expensive as it needs to go through the content of
+    /// every valid code cell.
+    fn build_index(&self) -> JupyterIndex {
+        let mut row_to_cell = vec![0];
+        let mut row_to_row_in_cell = vec![0];
+
+        for &pos in &self.valid_code_cells {
+            let line_count = match &self.raw.cells[pos as usize].source {
                 SourceValue::String(string) => {
-                    // TODO(konstin): is or isn't there a trailing newline per cell?
-                    // i've only seen these as array and never as string
-                    let line_count = u32::try_from(string.lines().count()).unwrap();
-                    jupyter_index.row_to_cell.extend(
-                        iter::repeat(u32::try_from(pos + 1).unwrap()).take(line_count as usize),
-                    );
-                    jupyter_index.row_to_row_in_cell.extend(1..=line_count);
-                    string.clone()
+                    if string.is_empty() {
+                        1
+                    } else {
+                        u32::try_from(NewlineWithTrailingNewline::from(string).count()).unwrap()
+                    }
                 }
                 SourceValue::StringArray(string_array) => {
-                    jupyter_index.row_to_cell.extend(
-                        iter::repeat(u32::try_from(pos + 1).unwrap()).take(string_array.len()),
-                    );
-                    jupyter_index
-                        .row_to_row_in_cell
-                        .extend(1..=u32::try_from(string_array.len()).unwrap());
-                    // lines already end in a newline character
-                    string_array.join("")
+                    if string_array.is_empty() {
+                        1
+                    } else {
+                        let trailing_newline =
+                            usize::from(string_array.last().map_or(false, |s| s.ends_with('\n')));
+                        u32::try_from(string_array.len() + trailing_newline).unwrap()
+                    }
                 }
             };
-            contents.push(cell_contents);
+            row_to_cell.extend(iter::repeat(pos + 1).take(line_count as usize));
+            row_to_row_in_cell.extend(1..=line_count);
+        }
+
+        JupyterIndex {
+            row_to_cell,
+            row_to_row_in_cell,
         }
-        // The last line doesn't end in a newline character
-        (contents.join("\n"), jupyter_index)
+    }
+
+    /// Return the notebook content.
+    ///
+    /// This is the concatenation of all Python code cells.
+    pub fn content(&self) -> &str {
+        &self.content
+    }
+
+    /// Return the Jupyter notebook index.
+    ///
+    /// The index is built only once when required. This is only used to
+    /// report diagnostics, so by that time all of the autofixes must have
+    /// been applied if `--fix` was passed.
```

Comment:
I'm loving all the documentation you're adding as you go. This is a very, very good habit. Props.